gen3/bin/replicate.sh

```diff
@@ -1,205 +1,266 @@
-    gen3_aws_run aws iam put-role-policy --role-name batch-operations-role --policy-name batch-operations-policy --policy-document "${policy//$'\n'}" --profile $profile
+  # the if run from source should take priority in the if statement becuase it implies there is a destination account and not all run in the same account
+  if [[ ! -z $runFromSource ]]; then
+    profileWithRole=$sourceProfile
+    profileWithSourceBucket=$sourceProfile
+    profileWithDestinationBucket=$destinationProfile
+  # if destination is set but you are not running from source account then you will be running from destination account with the source bucke in either adminvm account if sourceprofile not set or in specified source profile
+  elif [[ ! -z $destinationProfile ]]; then
+    profileWithRole=$destinationProfile
+    profileWithSourceBucket=$sourceProfile
+    profileWithDestinationBucket=$destinationProfile
+  # If these are not set then assume everything is run from sourceProfile, sourceProfile defaults to default profile (adminvm profile) if not set
   else
-    if [[ -z $(gen3_aws_run aws iam list-roles | jq -r .Roles[].RoleName | grep batch-operations-role) ]]; then
-      gen3_log_info "Creating batch-operations role"
-      local trustRelationship="{\"Version\": \"2012-10-17\",\"Statement\": [{\"Effect\": \"Allow\",\"Principal\": {\"Service\": \"batchoperations.s3.amazonaws.com\"},\"Action\": \"sts:AssumeRole\"}]}"
-      gen3_aws_run aws iam create-role --role-name batch-operations-role --assume-role-policy-document "${trustRelationship//$'\n'}"
-    fi
-    if [[ ! -z $(gen3_aws_run aws iam list-role-policies --role-name batch-operations-role | jq -r .PolicyNames[]) ]]; then
-      gen3_log_info "Old policy exists. Modifying for new policy"
-      gen3_aws_run aws iam delete-role-policy --role-name batch-operations-role --policy-name batch-operations-policy
+    profileWithRole=$sourceProfile
+    profileWithSourceBucket=$sourceProfile
+    profileWithDestinationBucket=$sourceProfile
+  fi
+  # Verify profile can reach buckets
+  gen3_replicate_verify_bucket_access
+  # Check if batch operations role exists. Create if it doesn't.
+  if [[ -z $(aws iam list-roles --profile $profileWithRole | jq -r .Roles[].RoleName | grep batch-operations-role) ]]; then
+    gen3_log_info "Creating batch operations role using profile $profileWithRole"
+    local trustRelationship="{\"Version\": \"2012-10-17\",\"Statement\": [{\"Effect\": \"Allow\",\"Principal\": {\"Service\": \"batchoperations.s3.amazonaws.com\"},\"Action\": \"sts:AssumeRole\"}]}"
+    aws iam create-role --role-name batch-operations-role --assume-role-policy-document "$trustRelationship"  --profile $profileWithRole
+  fi
+  # If the role is not created after that it implies the profile does not have permission to create the role and the process should be stopped.
+  if [[ -z $(aws iam list-roles --profile $profileWithRole | jq -r .Roles[].RoleName | grep batch-operations-role) ]]; then
```

Comment:
Sometime it takes for a while for aws really fulfill to create resources. The next code may returns error since the resource has not existed yet. You can put a loop here to get sleep until the resource is created. However we will see the loop everywhere. Probably the best solution is put a single loop just before the main part( launching job) to check if all the resources are ready?